fastify.d.ts

```diff
@@ -7,17 +7,21 @@ import * as http2 from 'http2';
 import * as https from 'https';
 import * as pino from 'pino';
 
-declare function fastify<HttpServer, HttpRequest, HttpResponse>(opts?: fastify.ServerOptions): fastify.FastifyInstance<HttpServer, HttpRequest, HttpResponse>;
+declare function fastify<
+  HttpServer extends http.Server = http.Server, 
+  HttpRequest extends http.IncomingMessage = http.IncomingMessage, 
+  HttpResponse extends http.ServerResponse = http.ServerResponse
+>(opts?: fastify.ServerOptions): fastify.FastifyInstance<HttpServer, HttpRequest, HttpResponse>;
 declare function fastify(opts?: fastify.ServerOptionsAsHttp): fastify.FastifyInstance<http.Server, http.IncomingMessage, http.ServerResponse>;
 declare function fastify(opts?: fastify.ServerOptionsAsSecureHttp): fastify.FastifyInstance<https.Server, http.IncomingMessage, http.ServerResponse>;
 declare function fastify(opts?: fastify.ServerOptionsAsHttp2): fastify.FastifyInstance<http2.Http2Server, http2.Http2ServerRequest, http2.Http2ServerResponse>;
 declare function fastify(opts?: fastify.ServerOptionsAsSecureHttp2): fastify.FastifyInstance<http2.Http2SecureServer, http2.Http2ServerRequest, http2.Http2ServerResponse>;
 
 declare namespace fastify {
 
-  type Plugin<HttpServer, HttpRequest, HttpResponse, T> = (instance: FastifyInstance<HttpServer, HttpRequest, HttpResponse>, opts: T, callback?: (err?: Error) => void) => void
+  type Plugin<HttpServer, HttpRequest, HttpResponse, T> = (instance: FastifyInstance<HttpServer, HttpRequest, HttpResponse>, opts: T, callback: (err?: Error) => void) => void
```

Comment:
the callback is optional. If you are not using it, you are supposed to return a promise (or declare it as an async function).

Reply:
my understanding is that using the callback is optional but the callback itself is always passed to the plugin/middleware.

In any case I'll try add an overload omitting the callback if you return a promise.

Reply:
Can both middleware and plugins return promises?
I can't seem to find docs regarding returning promises.

Reply:
plugins and hooks and handlers can return a promise or support async/await

Reply:
ok, so first off there is no special support for async/await. async/await is just syntactical sugar over promises. From the perspective of the type system async function are the same as regular functions which return a promise.

As for adding the overload, I can add an overload but there doesn't seem to be any gain in doing so.

The current type looks something like
```ts
cb(req: HttpRequest, res: HttpResponse, callback: (err?: Error) => void) => void
```
Typescript does not enforce that a function use all params that its passed.
In addition when the return type is set to void TS ignore the return value as well.
What this means is that all these are valid for the current definition that gets passed a callback.
```ts
server.use(() => {})
server.use((req) => {})
server.use((req, res) => {})
server.use((req, res, next) => {})
server.use(() => {return anything})
server.use((req) => {return anything})
server.use((req, res) => {return anything})
server.use((req, res, next) => {return anything})
server.use(async () => {})
server.use(async (req) => {})
server.use(async (req, res) => {})
server.use(async (req, res, next) => {})
server.use(async () => {return anything})
server.use(async (req) => {return anything})
server.use(async (req, res) => {return anything})
server.use(async (req, res, next) => {return anything})
```

If we were to add an additional overload that looked like
```ts
cb(req: HttpRequest, res: HttpResponse) => Promise<void>
```
It would not protect us from returning a promise and taking in next.
Reason being that since the type signature of the callback variant allows for functions which return promise (because of the `void` return type), as soon as we accept a 3rd param to our function it will just switch to using the callback signature and will not complain about using next.

Reply:
use does not support promises. register does.

Reply:
what would be good to support is that, if you are not receiving next, you have to return a promise.

Reply:
I would prefer to leave that as optional.

Reply:
I think all that's needed is to add `| Promise` to the very end here.